server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,296 @@
+
+    // Flexible versioning on ListOffsets, WriteTxnMarkers and OffsetsForLeaderEpoch. Also adds topic IDs (KIP-516)
+    IBP_2_8_IV0(-1, "2.8", "IV0"),
+
+    // Introduced topic IDs to LeaderAndIsr and UpdateMetadata requests/responses (KIP-516)
+    IBP_2_8_IV1(-1, "2.8", "IV1"),
+
+    // Introduce AllocateProducerIds (KIP-730)
+    IBP_3_0_IV0(1, "3.0", "IV0"),
+
+    // Introduce ListOffsets V7 which supports listing offsets by max timestamp (KIP-734)
+    // Assume message format version is 3.0 (KIP-724)
+    IBP_3_0_IV1(2, "3.0", "IV1"),
+
+    // Adds topic IDs to Fetch requests/responses (KIP-516)
+    IBP_3_1_IV0(3, "3.1", "IV0"),
+
+    // Support for leader recovery for unclean leader election (KIP-704)
+    IBP_3_2_IV0(4, "3.2", "IV0"),
+
+    IBP_3_3_IV0(5, "3.3", "IV0");
+
+    private final Optional<Short> metadataVersion;
+    private final String shortVersion;
+    private final String version;
+
+    MetadataVersion(int metadataVersion, String shortVersion, String subVersion) {
+        if (metadataVersion > 0) {
+            this.metadataVersion = Optional.of((short) metadataVersion);
+        } else {
+            this.metadataVersion = Optional.empty();
+        }
+        this.shortVersion = shortVersion;
+        if (subVersion.equals("")) {
+            this.version = shortVersion;
+        } else {
+            this.version = String.format("%s-%s", shortVersion, subVersion);
+        }
+    }
+
+    public Optional<Short> metadataVersion() {
+        return metadataVersion;
+    }
+
+    public boolean isSaslInterBrokerHandshakeRequestEnabled() {
+        return this.isAtLeast(IBP_0_10_0_IV1);
+    }
+
+    public boolean isOffsetForLeaderEpochSupported() {
+        return this.isAtLeast(IBP_0_11_0_IV2);
+    }
+
+    public boolean isFeatureVersioningSupported() {
+        return this.isAtLeast(IBP_2_7_IV0);
+    }
+
+    public boolean isTruncationOnFetchSupported() {
+        return this.isAtLeast(IBP_2_7_IV1);
+    }
+
+    public boolean isAlterIsrSupported() {
+        return this.isAtLeast(IBP_2_7_IV2);
+    }
+
+    public boolean isTopicIdsSupported() {
+        return this.isAtLeast(IBP_2_8_IV0);
+    }
+
+    public boolean isAllocateProducerIdsSupported() {
+        return this.isAtLeast(IBP_3_0_IV0);
+    }
+
+
+    public RecordVersion recordVersion() {
+        if (this.isLessThan(IBP_0_10_0_IV0)) {
+            return RecordVersion.V0;
+        } else if (this.isLessThan(IBP_0_11_0_IV0)) {
+            return RecordVersion.V1;
+        } else {
+            return RecordVersion.V2;
+        }
+    }
+
+    private static final Map<String, MetadataVersion> IBP_VERSIONS;
+    static {
+        {
+            IBP_VERSIONS = new HashMap<>();
+            Map<String, MetadataVersion> maxInterVersion = new HashMap<>();
+            for (MetadataVersion metadataVersion : MetadataVersion.values()) {
+                maxInterVersion.compute(metadataVersion.shortVersion, (__, currentMetadataVersion) -> {
```

Comment:
Since this loop goes through the enum values in order, we don't have to do anything fancy here... we can just unconditionally overwrite the previous mapping and it will be OK, since we're handling enum N+1 and the old enum was N

Reply:
Makes sense, I guess we'd just have a few extraneous puts which are overwritten.